File: cpp/3047_largest_square_area/largest_square_area.cpp
#include <cassert>
#include <tuple>
#include <vector>
using namespace std;

class Solution {
 public:
  long long largestSquareArea(vector<vector<int>>& bottomLeft, vector<vector<int>>& topRight) {
    int n = bottomLeft.size();
    int maxSide = 0;

    for (int i = 0; i < n; i++) {
      for (int j = i + 1; j < n; j++) {
        int w = min(topRight[i][0], topRight[j][0]) - max(bottomLeft[i][0], bottomLeft[j][0]);
        int h = min(topRight[i][1], topRight[j][1]) - max(bottomLeft[i][1], bottomLeft[j][1]);

        maxSide = max(maxSide, min(w, h));
      }
    }

    return 1LL * maxSide * maxSide;
  }
};

int main() {
  vector<tuple<vector<vector<int>>, vector<vector<int>>, long long>> tests{
      {{{1, 1}, {2, 2}, {3, 1}}, {{3, 3}, {4, 4}, {6, 6}}, 1},
      {{{1, 1}, {2, 2}, {1, 2}}, {{3, 3}, {4, 4}, {3, 4}}, 1},
      {{{1, 1}, {3, 3}, {3, 1}}, {{2, 2}, {4, 4}, {4, 2}}, 0},
  };

  for (auto [bottomLeft, topRight, expected] : tests) {
    assert(Solution().largestSquareArea(bottomLeft, topRight) == expected);
  }
}

File: rust/3047_largest_square_area/largest_square_area.rs
struct Solution;

impl Solution {
    pub fn largest_square_area(bottom_left: Vec<Vec<i32>>, top_right: Vec<Vec<i32>>) -> i64 {
        use std::cmp::{max, min};
        let n = bottom_left.len();
        let mut max_side = 0;

        for i in 0..n {
            for j in i + 1..n {
                let w = min(top_right[i][0], top_right[j][0])
                    - max(bottom_left[i][0], bottom_left[j][0]);
                let h = min(top_right[i][1], top_right[j][1])
                    - max(bottom_left[i][1], bottom_left[j][1]);

                max_side = max(max_side, min(w, h));
            }
        }

        (max_side as i64) * (max_side as i64)
    }
}

fn main() {
    let tests = vec![
        (
            vec![vec![1, 1], vec![2, 2], vec![3, 1]],
            vec![vec![3, 3], vec![4, 4], vec![6, 6]],
            1,
        ),
        (
            vec![vec![1, 1], vec![2, 2], vec![1, 2]],
            vec![vec![3, 3], vec![4, 4], vec![3, 4]],
            1,
        ),
        (
            vec![vec![1, 1], vec![3, 3], vec![3, 1]],
            vec![vec![2, 2], vec![4, 4], vec![4, 2]],
            0,
        ),
    ];

    for (bottom_left, top_right, expected) in tests {
        assert_eq!(
            Solution::largest_square_area(bottom_left, top_right),
            expected
        );
    }
}
